### PR TITLE
[compiler] Improve handling of refs

### PR DESCRIPTION
**Pull request by mvitousek**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #30917
* __->__ #30902
* #30912

Summary:
This change expands our handling of refs to build an understanding of nested refs within objects and functions that may return refs. It builds a special-purpose type system within the ref analysis that gives a very lightweight structural type to objects and array expressions (merging the types of all their members), and then propagating those types throughout the analysis (e.g., if `ref` has type `Ref`, then `{ x: ref }` and `[ref]` have type `Structural(value=Ref)` and `{x: ref}.anything` and `[ref][anything]` have type `Ref`).

This allows us to support structures that contain refs, and functions that operate over them, being created and passed around during rendering without at runtime accessing a ref value.

The analysis here uses a fixpoint to allow types to be fully propagated through the system, and we defend against diverging by widening the type of a variable if it could grow infinitely: so, in something like
```
let x = ref;
while (condition) {
  x = [x]
}
```
we end up giving `x` the type `Structural(value=Ref)`.

### CODE REVIEW

**Comment by mvitousek** [comment]
[Internal sync](https://www.internalfb.com/intern/everpaste/?handle=GNO_VhsuNpbDRjsBANqxKvVSnrYcbsIXAAAB) results

A few runtime errors that need to be addressed before this can land but a significant reduction in bailouts

**Comment by mvitousek** [comment]
Updated [sync results](https://www.internalfb.com/intern/everpaste/?handle=GHCIVhs4Dt0R4rscAA2Vr6zOrqUObsIXAAAB) from the top of this stack at #30917 (including some unrelated PRs), looks good

File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts [hunk 1]
```diff
@@ -42,319 +44,443 @@ import {isEffectHook} from './ValidateMemoizedEffectDependencies';
...
+  resetChanged(): void {
+    this.#changed = false;
+  }
+
+  hasChanged(): boolean {
+    return this.#changed;
+  }
+
+  override set(key: IdentifierId, value: RefAccessType): this {
+    function tyEqual(a: RefAccessType, b: RefAccessType): boolean {
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts:69]: nit: extract this into a static function?

File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts [hunk 2]
```diff
@@ -42,319 +44,443 @@ import {isEffectHook} from './ValidateMemoizedEffectDependencies';
...
+    return {kind: 'RefValue'};
+  } else if (isUseRefType(identifier)) {
+    return {kind: 'Ref'};
+  } else {
+    return {kind: 'None'};
+  }
+}
+
+function joinRefTypes(...types: Array<RefAccessType>): RefAccessType {
+  function joinRefRefTypes(
```
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts:134]: naming is hard

File: compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts (hunk 1 above)
Comment (josephsavona) [inline compiler/packages/babel-plugin-react-compiler/src/Validation/ValidateNoRefAccesInRender.ts:69]: extract into a static function?

**Review by josephsavona — APPROVED** [review]
so good

**Pull request merged**